File: python/hsfs/core/training_dataset_bundle.py
#
#   Copyright 2022 Logical Clocks AB
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.
#


class TrainingDatasetBundle:
    def __init__(
        self,
        version,
        training_dataset=None,
        training_dataset_splits=None,
        train_split=None,
        job=None,
    ):
        self._version = version
        self._training_dataset = training_dataset
        self._training_dataset_splits = training_dataset_splits
        self._train_split = train_split
        self._job = job

    def get_dataset(self, split=None):
        if not self._job:
            if split:
                return self._training_dataset_splits.get(split)
            else:
                if self._train_split and self._training_dataset_splits:
                    return self._training_dataset_splits.get(self._train_split)
                else:
                    return self._training_dataset

    def get_split_names(self):
        if not self._job:
            return list(self._training_dataset_splits.keys())

    def get_train_split_name(self):
        return self._train_split

    def get_job(self):
        return self._job

    @property
    def version(self):
        return self._version

File: java/spark/src/main/java/com/logicalclocks/hsfs/spark/TrainingDatasetBundle.java
/*
 *  Copyright (c) 2022-2023. Hopsworks AB
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 *  See the License for the specific language governing permissions and limitations under the License.
 *
 */

package com.logicalclocks.hsfs.spark;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.google.common.collect.Lists;
import com.logicalclocks.hsfs.spark.engine.SparkEngine;
import com.logicalclocks.hsfs.Split;
import lombok.Getter;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;

import java.util.List;
import java.util.Map;

public class TrainingDatasetBundle {

  @Getter
  private Integer version;
  private Map<String, Dataset<Row>> datasetSplits;
  private Dataset<Row> dataset;
  private List<String> labels;
  private boolean hasSplit = false;
  private Boolean inMemory = true;

  public TrainingDatasetBundle(Integer version, Dataset<Row> dataset, List<String> labels) {
    this.version = version;
    this.dataset = dataset;
    this.labels = labels;
  }

  public TrainingDatasetBundle(Integer version) {
    this.version = version;
    this.inMemory = false;
  }

  public TrainingDatasetBundle(Integer version, Map<String, Dataset<Row>> datasetSplits, List<String> labels) {
    this.version = version;
    this.datasetSplits = datasetSplits;
    this.labels = labels;
    this.hasSplit = true;
  }

  @JsonIgnore
  public List<Dataset<Row>> getDataset(Boolean splitLabels) {
    if (inMemory) {
      if (hasSplit) {
        return getDataset(Split.TRAIN, splitLabels);
      } else {
        if (splitLabels) {
          return SparkEngine.splitLabels(dataset, labels);
        } else {
          return Lists.newArrayList(dataset);
        }
      }
    } else {
      return null;
    }
  }

  @JsonIgnore
  public List<Dataset<Row>> getDataset(String split, Boolean splitLabels) {
    if (inMemory) {
      if (splitLabels) {
        return SparkEngine.splitLabels(datasetSplits.get(split), labels);
      } else {
        return Lists.newArrayList(datasetSplits.get(split));
      }
    } else {
      return null;
    }
  }
}
